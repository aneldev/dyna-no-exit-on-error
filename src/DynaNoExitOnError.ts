import { dynaStringify } from "dyna-stringify";
import { isNode } from "./isNode";


export interface IDynaNoExitOnErrorConfig {
  onUncaughtException?: (error: any, origin: string, errorJson: any) => void;   // Occurred on exceptions
  onUncaughtRejection?: (error: any, origin: string, errorJson: any) => void;   // Occurred on exception in Promises
  onError?: (error: any, origin: string, errorJson: any) => void;               // Occurred on any of the above
}

export class DynaNoExitOnError {
  constructor(private readonly _config: IDynaNoExitOnErrorConfig) {
    if (Object.keys(this._config).length === 0) console.warn("DynaNoExitOnError called with empty configuration, all uncaught exceptions and rejections will be swallowed! Help how to configure it: http://github.com/aneldev/dyna-no-exit-on-error");
    this.enable();
  }

  public enable(): void {
    if (!isNode) return;
    process.on('uncaughtException', this._handleUncaughtException);
    process.on('unhandledRejection', this._handleUncaughtRejection);
  }

  public disable(): void {
    if (!isNode) return;
    process.off('uncaughtException', this._handleUncaughtException);
    process.off('unhandledRejection', this._handleUncaughtRejection);
  }

  public dispose(): void {
    this.disable();
  }

  private _handleUncaughtException = (err: any, origin: string): void => {
    this._triggerError(err, origin);
  };

  private _handleUncaughtRejection = (reason: any, promise: Promise<any>): void => {
    promise; // 4ts
    this._triggerError(reason, 'Promise');
  };

  private _triggerError(error: any, origin: string): void {
    const {
      onUncaughtException,
      onUncaughtRejection,
      onError,
    } = this._config;

    const errorJson = this._buildErrorJson(error, origin);

    onUncaughtException && onUncaughtException(error, origin, errorJson);
    onUncaughtRejection && onUncaughtRejection(error, origin, errorJson);
    onError && onError(error, origin, errorJson);
  }

  private _buildErrorJson(error: any, origin: string): any {
    const dynaNoExitOnErrorInfo = {
      origin,
    };
    const t1 = dynaStringify(error);
    if (t1 !== '{}') return {
      ...JSON.parse(dynaStringify(error)),
      dynaNoExitOnErrorInfo,
    };

    return {
      name: error.name,
      message: error.message,
      stack: error.stack,
      hasOwnProperty: error.hasOwnProperty && error.hasOwnProperty(),
      fullMessage: error.toString(),
      dynaNoExitOnErrorInfo,
    };
  }
}
