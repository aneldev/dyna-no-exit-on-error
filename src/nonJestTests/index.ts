import "dyna-node-console";

// insert here only the NON JEST tests
import "./DynaNoExitOnError.plain";
