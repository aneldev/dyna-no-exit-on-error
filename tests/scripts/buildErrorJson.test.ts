import "jest";
import { DynaNoExitOnError } from "../../src";

describe('buildErrorJSON', () => {
  it('builds js native errors', async () => {
      try {
        (process as any).__notExistProperty.name;
      } catch (e) {
        expect(DynaNoExitOnError._buildErrorJson(e, 'origin')).toMatchSnapshot(true);
      }
  });
  it('builds new Error', async () => {
    expect(DynaNoExitOnError._buildErrorJson(new Error('Custom new error'), 'origin')).toMatchSnapshot(true);
  });
  it('builds object Errors', async () => {
    expect(DynaNoExitOnError._buildErrorJson({message: 'Custom object error'}, 'origin')).toMatchSnapshot(true);
  });
});

