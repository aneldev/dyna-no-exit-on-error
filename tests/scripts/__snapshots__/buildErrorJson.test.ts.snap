// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildErrorJSON builds js native errors 1`] = `
Object {
  "dynaNoExitOnErrorInfo": Object {
    "origin": "origin",
  },
  "fullMessage": "TypeError: Cannot read property 'name' of undefined",
  "hasOwnProperty": false,
  "message": "Cannot read property 'name' of undefined",
  "name": "TypeError",
  "stack": "TypeError: Cannot read property 'name' of undefined
    at /Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:7:45
    at step (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:33:23)
    at Object.next (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:14:53)
    at /Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:8:71
    at new Promise (<anonymous>)
    at Object.<anonymous>.__awaiter (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:4:12)
    at Object.<anonymous> (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:5:33)
    at Object.asyncJestTest (/Users/dennis/dev/dyna/dyna-no-exit-on-error/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)
    at /Users/dennis/dev/dyna/dyna-no-exit-on-error/node_modules/jest-jasmine2/build/queueRunner.js:43:12
    at new Promise (<anonymous>)",
}
`;

exports[`buildErrorJSON builds new Error 1`] = `
Object {
  "dynaNoExitOnErrorInfo": Object {
    "origin": "origin",
  },
  "fullMessage": "Error: Custom new error",
  "hasOwnProperty": false,
  "message": "Custom new error",
  "name": "Error",
  "stack": "Error: Custom new error
    at /Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:13:46
    at step (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:33:23)
    at Object.next (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:14:53)
    at /Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:8:71
    at new Promise (<anonymous>)
    at Object.<anonymous>.__awaiter (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:4:12)
    at Object.<anonymous> (/Users/dennis/dev/dyna/dyna-no-exit-on-error/tests/scripts/buildErrorJson.test.ts:12:26)
    at Object.asyncJestTest (/Users/dennis/dev/dyna/dyna-no-exit-on-error/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)
    at /Users/dennis/dev/dyna/dyna-no-exit-on-error/node_modules/jest-jasmine2/build/queueRunner.js:43:12
    at new Promise (<anonymous>)",
}
`;

exports[`buildErrorJSON builds object Errors 1`] = `
Object {
  "dynaNoExitOnErrorInfo": Object {
    "origin": "origin",
  },
  "message": "Custom object error",
}
`;
